cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -155,47 +212,97 @@ ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile,
         "Couldn't unzip artifact %s, please check it is a valid jar file.", artifactId), e);
     }
 
-    if (mainClassName != null) {
-      try (CloseableClassLoader artifactClassLoader = artifactClassLoaderFactory.createClassLoader(artifactLocation)) {
-        Object appMain = artifactClassLoader.loadClass(mainClassName).newInstance();
-        if (!(appMain instanceof Application)) {
-          // we don't want to error here, just don't record an application class.
-          // possible for 3rd party plugin artifacts to have the main class set
-          return builder;
-        }
+    if (mainClassName == null) {
+      return builder;
+    }
 
-        Application app = (Application) appMain;
+    try (CloseableClassLoader artifactClassLoader = artifactClassLoaderFactory.createClassLoader(unpackedDir)) {
+      extractDatasets(unpackedDir, artifactClassLoader, builder);
 
-        java.lang.reflect.Type configType;
-        // if the user parameterized their application, like 'xyz extends Application<T>',
-        // we can deserialize the config into that object. Otherwise it'll just be a Config
-        try {
-          configType = Artifacts.getConfigType(app.getClass());
-        } catch (Exception e) {
-          throw new InvalidArtifactException(String.format(
-            "Could not resolve config type for Application class %s in artifact %s. " +
-              "The type must extend Config and cannot be parameterized.", mainClassName, artifactId));
-        }
+      Object appMain = artifactClassLoader.loadClass(mainClassName).newInstance();
```

Comment:
It can be rewritten to use bytecode inspection as well, but probably would involve quite a lot of new code, mainly for resolving parameterized type recursively along the classes inheritance hierarchy.